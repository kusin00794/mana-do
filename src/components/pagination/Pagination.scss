@import '../../styles/variables.scss';

.pagination {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: flex-end;
  margin: 15px 0 0;

  .pagination-item {
    padding: 8px 16px;
    border: 1px solid $pagination-border;
    text-decoration: none;
    color: $black-color;
    transition: background-color 0.3s;

    &.disabled {
      background-color: $pagination-disabled-bg-color;
      cursor: default;

      a {
        cursor: default;
      }
    }

    &.active {
      border: 1px solid $pagination-bg-color;
      background-color: $pagination-bg-color;
      color: $white-color;
      cursor: default;

      a {
        color: $white-color;
        cursor: default;
      }
    }

    &:hover {
      cursor: pointer;
    }

    &:hover:not(.active) {
      background-color: $pagination-border;
    }
  }
}
