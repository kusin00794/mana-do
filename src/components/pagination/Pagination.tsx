import React, { FC, useState, useEffect } from 'react';
import { Link } from 'react-router-dom';

import { IPagination, IPaginationInfo } from './Pagination.d';

import './Pagination.scss';

export const defaultPerPage = 10;
export const firstPage = 1;

export const defaultPaginationInfo: IPaginationInfo = { perPage: defaultPerPage, currentPage: firstPage };

const Pagination: FC<IPagination> = ({ total, perPage, handlePagination, externalCurrentPage }) => {
  const [currentPage, setCurrentPage] = useState<number>(externalCurrentPage || firstPage);
  const finalPerPage = perPage || defaultPerPage;
  const pageNumbers = [];
  for (let i = 1; i <= Math.ceil(total / finalPerPage); i++) {
    pageNumbers.push(i);
  }

  useEffect(() => {
    externalCurrentPage && setCurrentPage(externalCurrentPage);
  }, [externalCurrentPage]);

  const handleOnClick = (e: React.MouseEvent, page: number) => {
    e.preventDefault();
    if (currentPage === page) {
      return;
    }

    setCurrentPage(page);
    handlePagination({
      currentPage: page,
      perPage: finalPerPage,
    });
  };

  const handleOnClickNext = (e: React.MouseEvent) => {
    if (currentPage >= pageNumbers.length) {
      return;
    }

    e.preventDefault();
    const newCurrentPage = currentPage + 1;
    setCurrentPage(currentPage + 1);
    handlePagination({
      currentPage: newCurrentPage,
      perPage: finalPerPage,
    });
  };

  const handleOnClickPrev = (e: React.MouseEvent) => {
    if (currentPage <= firstPage) {
      return;
    }

    e.preventDefault();
    const newCurrentPage = currentPage - 1;
    setCurrentPage(currentPage - 1);
    handlePagination({
      currentPage: newCurrentPage,
      perPage: finalPerPage,
    });
  };

  return (
    <ul className="pagination">
      <li
        className={`pagination-prev pagination-item${currentPage <= firstPage ? ' disabled' : ''}`}
        onClick={handleOnClickPrev}
      >
        <Link to="#">
          <i className="fa fa-fw fa-angle-left" />
        </Link>
      </li>
      {pageNumbers.map((page, index) => {
        return (
          <li
            key={`pagination-item-${index}`}
            className={`pagination-page-${page} pagination-item${page === currentPage ? ' active' : ''}`}
            onClick={(e) => handleOnClick(e, page)}
          >
            <Link to="#">{page}</Link>
          </li>
        );
      })}
      <li
        className={`pagination-next pagination-item${currentPage >= pageNumbers.length ? ' disabled' : ''}`}
        onClick={handleOnClickNext}
      >
        <Link to="#">
          <i className="fa fa-fw fa-angle-right" />
        </Link>
      </li>
    </ul>
  );
};

export default Pagination;
