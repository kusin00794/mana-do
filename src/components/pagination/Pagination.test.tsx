import * as React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, fireEvent, wait } from '@testing-library/react';

import { testSnapshots } from '../../utils/test';
import Pagination from './Pagination';

const Wrapper = (props: any) => (
  <BrowserRouter>
    <Pagination {...props} />
  </BrowserRouter>
);

const defaultProps = {
  total: 21,
  handlePagination: jest.fn(),
};

describe('Pagination Page', () => {
  testSnapshots(Wrapper, [
    {
      props: {},
      description: 'render Pagination Page',
    },
  ]);
});

describe('Pagination logic', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.restoreAllMocks();
    jest.resetModules();
  });

  test('Should render correct prev button', async () => {
    const wrapper = render(<Wrapper {...defaultProps} />);
    const { container } = wrapper;
    const prevBtn = container.querySelector('.pagination-prev');
    expect(prevBtn).toHaveClass('disabled');
    await wait(() => {
      if (prevBtn) {
        fireEvent.click(prevBtn);
      }
    });
    expect(defaultProps.handlePagination).not.toHaveBeenCalled();
  });

  test('Should render correct next button', async () => {
    const wrapper = render(<Wrapper {...defaultProps} externalCurrentPage={3} />);
    const { container } = wrapper;
    const nextBtn = container.querySelector('.pagination-next');
    expect(nextBtn).toHaveClass('disabled');
    await wait(() => {
      if (nextBtn) {
        fireEvent.click(nextBtn);
      }
    });
    expect(defaultProps.handlePagination).not.toHaveBeenCalled();
  });

  test('Should have correct behavior when click next button and prev', async () => {
    const wrapper = render(<Wrapper {...defaultProps} />);
    const { container, rerender } = wrapper;

    const nextBtn = container.querySelector('.pagination-next');
    const expectedNextValue = { currentPage: 2, perPage: 10 };
    await wait(() => {
      if (nextBtn) {
        fireEvent.click(nextBtn);
      }
    });
    expect(defaultProps.handlePagination).toHaveBeenCalledWith(expectedNextValue);

    rerender(<Wrapper {...defaultProps} />);

    const prevBtn = container.querySelector('.pagination-prev');
    const expectedPrevValue = { currentPage: 1, perPage: 10 };
    await wait(() => {
      if (prevBtn) {
        fireEvent.click(prevBtn);
      }
    });
    expect(defaultProps.handlePagination).toHaveBeenCalledWith(expectedPrevValue);
  });

  test('Should have correct behavior when click item', async () => {
    const wrapper = render(<Wrapper {...defaultProps} />);
    const { container } = wrapper;

    const firstPage = container.querySelector('.pagination-page-1');
    await wait(() => {
      if (firstPage) {
        fireEvent.click(firstPage);
      }
    });
    expect(defaultProps.handlePagination).not.toHaveBeenCalled();

    const lastPage = container.querySelector('.pagination-page-2');
    const expectedSecPageValue = { currentPage: 2, perPage: 10 };
    await wait(() => {
      if (lastPage) {
        fireEvent.click(lastPage);
      }
    });
    expect(defaultProps.handlePagination).toHaveBeenCalledWith(expectedSecPageValue);
  });
});
