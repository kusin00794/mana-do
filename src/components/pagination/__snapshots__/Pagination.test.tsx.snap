// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Page snapshots render Pagination Page 1`] = `
<ul
  className="pagination"
>
  <li
    className="pagination-prev pagination-item disabled"
    onClick={[Function]}
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <i
        className="fa fa-fw fa-angle-left"
      />
    </a>
  </li>
  <li
    className="pagination-next pagination-item disabled"
    onClick={[Function]}
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <i
        className="fa fa-fw fa-angle-right"
      />
    </a>
  </li>
</ul>
`;
